### PR TITLE
refactor(binder): gracefully `pop_context` for rcte's binding; separate validation & actual binding process

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

the main reason is to gracefully `pop_context` even there are errors during binding phase, same as `bind_query`.

the rest is just logically separate each phase.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/binder/query.rs [hunk 1]
```diff
@@ -289,51 +289,22 @@ impl Binder {
...
-                    op: SetOperator::Union,
-                    all,
-                    left,
-                    right,
-                } = body
+                ) = Self::validate_rcte(query)?
                 else {
                     return Err(ErrorCode::BindError(
-                        "`UNION` is required in recursive CTE".to_string(),
+                        "expect `SetOperation` as the return type of validation".into(),
```
Comment (TennyZhuang) [inline src/frontend/src/binder/query.rs:303]: This is a user-facing error, while `SetOperation` is a uncommon terminology to users, so prefer keeping `UNION` here.

BTW, actually only `UNION` is valid here, other set operations will be rejected.

**Review by TennyZhuang — APPROVED** [review]
Rest LGTM

File: src/frontend/src/binder/query.rs (hunk 1 above)
Comment (TennyZhuang) [reply src/frontend/src/binder/query.rs:303]: OK, I got it. It's an internal error.

Comment (xzhseh) [reply src/frontend/src/binder/query.rs:303]: yep and just for reference, the actual user-facing error is here: https://github.com/risingwavelabs/risingwave/pull/16458/files#diff-9defd28e358376355286b3fe37ff756a22c7bfd3bd5e25bb08848948dd2e09b5R373

**Pull request merged**